test/testffmpeg_vulkan.c

```diff
@@ -0,0 +1,973 @@
+            return -1;
+        }
+        context->commandBuffers = commandBuffers;
+
+        VkCommandBufferAllocateInfo commandBufferAllocateInfo = { 0 };
+        commandBufferAllocateInfo.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+        commandBufferAllocateInfo.commandPool = context->commandPool;
+        commandBufferAllocateInfo.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+        commandBufferAllocateInfo.commandBufferCount = needed;
+        VkResult result = context->vkAllocateCommandBuffers(context->device, &commandBufferAllocateInfo, &context->commandBuffers[context->commandBufferCount]);
+        if (result != VK_SUCCESS) {
+            SDL_SetError("vkAllocateCommandBuffers(): %s", getVulkanResultString(result));
+            return -1;
+        }
+
+        context->commandBufferCount = commandBufferCount;
+    }
+    return 0;
+}
+
+static int ScheduleFrameTransferToSDL(VulkanVideoContext *context, AVFrame *frame)
+{
+    AVVkFrame *pVkFrame = (AVVkFrame *)frame->data[0];
+
+    VkTimelineSemaphoreSubmitInfo timeline = { 0 };
+    timeline.sType = VK_STRUCTURE_TYPE_TIMELINE_SEMAPHORE_SUBMIT_INFO;
+    timeline.waitSemaphoreValueCount = 1;
+    timeline.pWaitSemaphoreValues = pVkFrame->sem_value;
+
+    VkPipelineStageFlags pipelineStageMask = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+    VkSubmitInfo submitInfo = { 0 };
+    submitInfo.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+    submitInfo.waitSemaphoreCount = 1;
+    submitInfo.pWaitSemaphores = pVkFrame->sem;
+    submitInfo.pWaitDstStageMask = &pipelineStageMask;
+    submitInfo.signalSemaphoreCount = 1;
+    submitInfo.pSignalSemaphores = &context->waitSemaphores[context->commandBufferIndex];
+    submitInfo.pNext = &timeline;
+
+    if (pVkFrame->layout[0] != VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL) {
+        VkCommandBuffer commandBuffer = context->commandBuffers[context->commandBufferIndex];
+
+        VkCommandBufferBeginInfo beginInfo = { 0 };
+        beginInfo.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+        beginInfo.flags = 0;
+        context->vkBeginCommandBuffer(commandBuffer, &beginInfo);
+
+        VkImageMemoryBarrier2 barrier = { 0 };
+        barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER_2;
+        barrier.srcAccessMask = VK_ACCESS_2_NONE;
+        barrier.dstAccessMask = VK_ACCESS_2_SHADER_SAMPLED_READ_BIT;
+        barrier.oldLayout = pVkFrame->layout[0];
+        barrier.newLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+        barrier.image = pVkFrame->img[0];
+        barrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+        barrier.subresourceRange.levelCount = 1;
+        barrier.subresourceRange.layerCount = 1;
+        barrier.srcQueueFamilyIndex = pVkFrame->queue_family[0];
+        barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+        barrier.srcStageMask = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+        barrier.dstStageMask = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
```

Comment:
I'm slightly surprised by the srcAccessMask and srcStageMask here.  If this is for the video decode, I suppose I would have expected something like:

`srcAccessMask = VK_ACCESS_2_VIDEO_DECODE_WRITE_BIT_KHR`
`srcStageMask = VK_PIPELINE_STAGE_2_VIDEO_DECODE_BIT_KHR`

Reply:
When both ffmpeg and Granite transition the image, they use an access mask of 0. When ffmpeg transitions the image, it uses VK_PIPELINE_STAGE_2_VIDEO_DECODE_BIT_KHR for both src and dst stage, and when Granite transitions the image, it uses VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT for both src and dst stage.

I'm not sure what the correct semantics are, but I can change it to use VK_PIPELINE_STAGE_2_VIDEO_DECODE_BIT_KHR  for the src stage, since we know that's where it is when it's coming from ffmpeg.

Reply:
Fixed!

Reply:
Actually, this gives a validation error:
VUID-vkCmdPipelineBarrier2-srcStageMask-03849(ERROR / SPEC): msgNum: -1477518705 - Validation Error: [ VUID-vkCmdPipelineBarrier2-srcStageMask-03849 ] Object 0: handle = 0x1be42d77af0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xa7eeda8f | vkCmdPipelineBarrier2(): pDependencyInfo->pImageMemoryBarriers[0].srcStageMask (VK_PIPELINE_STAGE_2_VIDEO_DECODE_BIT_KHR) is not compatible with the queue family properties (VK_QUEUE_GRAPHICS_BIT|VK_QUEUE_COMPUTE_BIT|VK_QUEUE_TRANSFER_BIT|VK_QUEUE_SPARSE_BINDING_BIT) of this command buffer. The Vulkan spec states: The srcStageMask member of any element of the pMemoryBarriers, pBufferMemoryBarriers, or pImageMemoryBarriers members of pDependencyInfo must only include pipeline stages valid for the queue family that was used to create the command pool that commandBuffer was allocated from (https://vulkan.lunarg.com/doc/view/1.3.275.0/windows/1.3-extensions/vkspec.html#VUID-vkCmdPipelineBarrier2-srcStageMask-03849)

Reply:
VUID-VkImageMemoryBarrier2-srcAccessMask-04859(ERROR / SPEC): msgNum: 1537468607 - Validation Error: [ VUID-VkImageMemoryBarrier2-srcAccessMask-04859 ] Object 0: handle = 0x1fb113b1c30, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x5ba3e8bf | vkCmdPipelineBarrier2(): pDependencyInfo->pImageMemoryBarriers[0].srcAccessMask (VK_ACCESS_2_VIDEO_DECODE_WRITE_BIT_KHR) is not supported by stage mask (VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT). The Vulkan spec states: If srcAccessMask includes VK_ACCESS_2_VIDEO_DECODE_WRITE_BIT_KHR, srcStageMask must include VK_PIPELINE_STAGE_2_VIDEO_DECODE_BIT_KHR (https://vulkan.lunarg.com/doc/view/1.3.275.0/windows/1.3-extensions/vkspec.html#VUID-VkImageMemoryBarrier2-srcAccessMask-04859)

Reply:
Ah, I see, so that barrier was on the ffmpeg video decode queue.  Yeah, ignore my suggestion here, was fine as is.